# Classification with _Random Forest_

In [1]:
import seaborn as sns
import sklearn as sk
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split #We need this to split the data

In [6]:
df = pd.read_csv('creditcard.csv')
#df = df[['Survived','Pclass', 'Age', 'SibSp', 'Sex', 'Parch']] #make this selection before you dropna
df = df.dropna() #first get rid of rows with empty cells 
df.head(30)


Time        V1        V2        V3        V4        V5        V6  \
0      0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1      0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2      1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3      1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4      2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
5      2 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728   
6      4  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708   
7      7 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118   
8      7 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818   
9      9 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761   
10    10  1.449044 -1.176339  0.913860 -1.375667 -1.971383 -0.629152   
11    10  0.384978  0.616109 -0.874300 -0.094019  2.924584  3.317027   
12    10  1.249999 -1.221637  0.383930 -1.234899 -1.485419 -0.753230   
13    11  1.069374  0.287722  0.828613  2.712520 -0.178398  0.337544   
14    12 -2.791855 -0.327771  1.641750  1.767473 -0.136588  0.807596   
15    12 -0.752417  0.345485  2.057323 -1.468643 -1.158394 -0.077850   
16    12  1.103215 -0.040296  1.267332  1.289091 -0.735997  0.288069   
17    13 -0.436905  0.918966  0.924591 -0.727219  0.915679 -0.127867   
18    14 -5.401258 -5.450148  1.186305  1.736239  3.049106 -1.763406   
19    15  1.492936 -1.029346  0.454795 -1.438026 -1.555434 -0.720961   
20    16  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   
21    17  0.962496  0.328461 -0.171479  2.109204  1.129566  1.696038   
22    18  1.166616  0.502120 -0.067300  2.261569  0.428804  0.089474   
23    18  0.247491  0.277666  1.185471 -0.092603 -1.314394 -0.150116   
24    22 -1.946525 -0.044901 -0.405570 -1.013057  2.941968  2.955053   
25    22 -2.074295 -0.121482  1.322021  0.410008  0.295198 -0.959537   
26    23  1.173285  0.353498  0.283905  1.133563 -0.172577 -0.916054   
27    23  1.322707 -0.174041  0.434555  0.576038 -0.836758 -0.831083   
28    23 -0.414289  0.905437  1.727453  1.473471  0.007443 -0.200331   
29    23  1.059387 -0.175319  1.266130  1.186110 -0.786002  0.578435   

          V7        V8        V9  ...       V21       V22       V23       V24  \
0   0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928   
1  -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846   
2   0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281   
3   0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575   
4   0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267   
5   0.476201  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427   
6  -0.005159  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055   
7   1.120631 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709   
8   0.370145  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592   
9   0.651583  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050   
10 -1.423236  0.048456 -1.720408  ... -0.009302  0.313894  0.027740  0.500512   
11  0.470455  0.538247 -0.558895  ...  0.049924  0.238422  0.009130  0.996710   
12 -0.689405 -0.227487 -2.094011  ... -0.231809 -0.483285  0.084668  0.392831   
13 -0.096717  0.115982 -0.221083  ... -0.036876  0.074412 -0.071407  0.104744   
14 -0.422911 -1.907107  0.755713  ...  1.151663  0.222182  1.020586  0.028317   
15 -0.608581  0.003603 -0.436167  ...  0.499625  1.353650 -0.256573 -0.065084   
16 -0.586057  0.189380  0.782333  ... -0.024612  0.196002  0.013802  0.103758   
17  0.707642  0.087962 -0.665271  ... -0.194796 -0.672638 -0.156858 -0.888386   
18 -1.559738  0.160842  1.233090  ... -0.503600  0.984460  2.458589  0.042119   
19 -1.080664 -0.053127 -1.978682  ... -0.177650 -0.175074  0.040002  0.295814   
20 -0.878586  0.445290 -0.446196  ... -0.295583 -0.571955 -0.050881 -0

In [7]:
df['Class'].value_counts()

0    99776
1      223
Name: Class, dtype: int64

In [9]:
from sklearn.preprocessing import normalize #get the function needed to normalize our data.

X = df.loc[:,'Time' : 'Amount'] #create the X matrix

y = df['Class'] #create the y-variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables
#the order of the above variables is very important
X_train.head()

Time        V1        V2        V3        V4        V5        V6  \
92336  63884  0.805625 -1.695688  1.352358 -0.282137 -2.253216  0.017385   
62015  50089  1.100031 -0.102983  1.111263  1.292381 -0.933713 -0.115089   
5005    4573 -1.552078 -0.531074  1.974243  1.013701  1.024195 -1.144346   
56848  47632 -0.419158  0.949003  1.336161  0.854330  0.442809  0.107296   
99373  67089  0.054913  1.758103 -2.796921  1.320415  1.549875 -0.875941   

             V7        V8        V9  ...       V20       V21       V22  \
92336 -1.217197  0.202760 -0.152576  ... -0.005321 -0.004854  0.042170   
62015 -0.540140  0.190514  0.607337  ... -0.162889  0.151511  0.592482   
5005   0.002698  0.048984  1.339099  ...  0.241350 -0.181249 -0.715027   
56848  0.772343 -0.182008 -0.640852  ...  0.015254  0.213674  0.779649   
99373  0.498296  0.236048 -0.683338  ...  0.008538 -0.125729 -0.211676   

            V23       V24       V25       V26       V27       V28  Amount  
92336 -0.132694  0.590563 -0.139888  1.096028 -0.039289  0.058267  239.76  
62015 -0.125180  0.561148  0.570113 -0.244571  0.051015  0.021777   15.93  
5005   0.328461  0.268026 -0.015884 -0.726971  0.026191  0.155041   83.72  
56848 -0.181551  0.038274 -0.493093 -0.334844 -0.222714 -0.062746   19.99  
99373  0.142032 -1.351474 -0.221678 -0.351459 -0.100930 -0.083579    8.49  

[5 rows x 30 columns]

In [10]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier() #create a Random forest-classifier 
rf = rf.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data
rf.score(X_test, y_test) #calculate the fit on the test data

/Users/marissa/opt/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9993666666666666

In [11]:
from sklearn.metrics import confusion_matrix
y_test_pred = rf.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix"
cm

array([[29930,     4],
       [   15,    51]])

In [12]:
conf_matrix = pd.DataFrame(cm, index=['Not fraud (actual)', 'Fraud (actual)'], columns = ['Not fraud (predicted)', 'Fraud (predicted)']) 
conf_matrix

Not fraud (predicted)  Fraud (predicted)
Not fraud (actual)                  29930                  4
Fraud (actual)                         15                 51

The way to read this is that of the fraud, 51 are correctly predicted as 'Fraud', and 4 that were predicted as fraud were actually not fraud. And of those who were predicted as not fraud, 29930 were predicted correctly, while 15 actually were fradu. The _recall_ and _precision_ for the fraud predictions:

$recall = \frac{51}{ 15 + 51} = .77$

$precision = \frac{51 }{ 4 + 51} = .92$



The model is pretty good at predicting what is not fraud. It did detect 4 items as fraud when it was not actually fraud but for a bank it is better to be precise than to be accurate, because they can call customers to verify purchases so it is better to be safe.

# Try with new Parameters

In [13]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=50) #create a Random forest-classifier 
rf = rf.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data
rf.score(X_test, y_test) #calculate the fit on the test data

0.9994666666666666

In [14]:
from sklearn.metrics import confusion_matrix
y_test_pred = rf.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix"
cm

array([[29934,     0],
       [   16,    50]])

In [15]:
conf_matrix = pd.DataFrame(cm, index=['Not fraud (actual)', 'Fraud (actual)'], columns = ['Not fraud (predicted)', 'Fraud (predicted)']) 
conf_matrix

Not fraud (predicted)  Fraud (predicted)
Not fraud (actual)                  29934                  0
Fraud (actual)                         16                 50

The way to read this is that of the fraud, 50 are correctly predicted as 'Fraud', and 0 that were predicted as fraud were actually not fraud. And of those that were predicted as not fraud, 29934 were predicted correctly, while 16 actually were fraud. The _recall_ and _precision_ for the fraud predictions:

$recall = \frac{50}{ 16 + 50} = .75$

$precision = \frac{50 }{ 0 + 50} = 1$



When using the new number of estimators (n_estimators=50 as opposed to the default of 100) we got an increase in _precision_- meaning all of the prediced fraud was actually fraud, this may seem good, but for a bank this is not the relationship we are most focused on.
With the new number of estimators, we got a decrease in _recall_ - meaning less of the real fraud was detected.

For a bank, recall is the most important relationship- we want to be able to detect all of the actual fraud. So in this case using the default number of estimators was better than n_estimators= 50